lib/cabal-project-parser.nix

```diff
@@ -95,10 +95,90 @@ let
         };
 
+  parseSourceRepositoryPackages = cabalProjectFileName: lookupSha256: source-repo-override: projectFile:
+    let
+      blocks = pkgs.lib.splitString "\nsource-repository-package\n" ("\n" + projectFile);
+      initialText = pkgs.lib.lists.take 1 blocks;
+      repoBlocks = builtins.map (parseBlock cabalProjectFileName lookupSha256) (pkgs.lib.lists.drop 1 blocks);
+      overrideSourceRepo = sourceRepo: (source-repo-override.${sourceRepo.url} or (pkgs.lib.id)) sourceRepo;
+    in {
+      sourceRepos = pkgs.lib.lists.map (block: overrideSourceRepo block.sourceRepo) repoBlocks;
+      otherText = pkgs.lib.strings.concatStringsSep "\n" (
+        initialText
+        ++ (builtins.map (x: x.otherText) repoBlocks));
+    };
+
+  # Parse a repository
+  parseRepositoryBlock = cabalProjectFileName: sha256map: cabal-install: nix-tools: block:
+    let
+      lines = pkgs.lib.splitString "\n" block;
+      x = span (pkgs.lib.strings.hasPrefix " ") (__tail lines);
+      attrs = parseBlockLines x.fst;
+    in rec {
+      name = __head lines;
+      repo = attrs;
+      home =
+        pkgs.evalPackages.runCommandLocal name {
+          nativeBuildInputs = [ cabal-install pkgs.evalPackages.curl nix-tools ];
+          LOCALE_ARCHIVE = pkgs.lib.optionalString (pkgs.evalPackages.stdenv.buildPlatform.libc == "glibc") "${pkgs.evalPackages.glibcLocales}/lib/locale/locale-archive";
+          LANG = "en_US.UTF-8";
+          outputHashMode = "recursive";
+          outputHashAlgo = "sha256";
+          outputHash = attrs."--sha256" or sha256map.${attrs.url};
+        } ''
+            mkdir -p $out/.cabal
+            cat <<EOF > $out/.cabal/config
+            repository ${name}
+              url: ${attrs.url}
+              ${pkgs.lib.optionalString (attrs ? secure) "secure: ${attrs.secure}"}
+              ${pkgs.lib.optionalString (attrs ? root-keys) "root-keys: ${attrs.root-keys}"}
+              ${pkgs.lib.optionalString (attrs ? key-threshold) "key-threshold: ${attrs.key-threshold}"}
+            EOF
+
+            export SSL_CERT_FILE=${pkgs.evalPackages.cacert}/etc/ssl/certs/ca-bundle.crt
+            mkdir -p $out/.cabal/packages/${name}
+            HOME=$out cabal new-update ${name}
+            ls -ld $out/.cabal
+        '';
+      hackage = import (
+        pkgs.evalPackages.runCommandLocal ("hackageg-to-nix-" + name) {
+          nativeBuildInputs = [ cabal-install pkgs.evalPackages.curl nix-tools ];
+          LOCALE_ARCHIVE = pkgs.lib.optionalString (pkgs.evalPackages.stdenv.buildPlatform.libc == "glibc") "${pkgs.evalPackages.glibcLocales}/lib/locale/locale-archive";
+          LANG = "en_US.UTF-8";
+        } ''
+            mkdir -p $out
+            hackage-to-nix $out ${home}/.cabal/packages/${name}/01-index.tar ${attrs.url}
+        '');
+      tarball = {
+        inherit name;
+        index = home + "/.cabal/packages/${name}/01-index.tar.gz";
+      };
+      otherText = ''
+        repository ${name}
```

Comment:
Again, this is inconsistent: for s-r-ps the `otherText` is just the leftover text and we construct the stanza which points to the constructed repo in `call-cabal-project-to-nix`.